### PR TITLE
Bug: Mutex locks not released on error

### LINKED ISSUE #4343: Transactions getting stuck in the "approved" state
Approved is the state just before "submitted", which should basically be invisible, because we should submit instantly and reliably. Many users have reported (and I have also experienced) a condition where transactions fall into the "Approved" state, and never progress from there.

This was a previous issue, but I can't find it, so I am creating this to track the issue in the meanwhile.

We should ensure that "approved" transactions are re-tried at least some number of times, and then marked as "failed". Other solutions are welcome.

Somewhat mitigated by #3771.

### PR DESCRIPTION
**Pull request by kumavis**
Fixes https://github.com/MetaMask/metamask-extension/issues/4343

Auditing and fixing use of locks:
- [x] MetamaskController
  - [x] createNewVaultAndKeychain
  - [x] createNewVaultAndRestore
- [x] PendingTxTracker
  - [x] _checkPendingTxs
- [x] NonceTracker
  - [x] getNonceLock
- [x] TransactionController
  - [x] approveTransaction

### CODE REVIEW

File: app/scripts/controllers/transactions/index.js [hunk 1]
```diff
@@ -264,7 +264,12 @@ class TransactionController extends EventEmitter {
       // must set transaction to submitted/failed before releasing lock
       nonceLock.releaseLock()
     } catch (err) {
-      this.txStateManager.setTxStatusFailed(txId, err)
+      // this is try-catch wrapped so that we can guarantee that the nonceLock is released
+      try {
+        this.txStateManager.setTxStatusFailed(txId, err)
+      } catch (err) {
+        console.error(err)
```
Comment (danfinlay) [inline app/scripts/controllers/transactions/index.js]: Should use `log.error()`

**Comment by kumavis** [comment]
@frankiebee you were working on this sort of thing before, so you might be interested in these changes

File: app/scripts/controllers/transactions/index.js [hunk 2]
```diff
@@ -264,7 +264,12 @@ class TransactionController extends EventEmitter {
       // must set transaction to submitted/failed before releasing lock
       nonceLock.releaseLock()
     } catch (err) {
-      this.txStateManager.setTxStatusFailed(txId, err)
+      // this is try-catch wrapped so that we can guarantee that the nonceLock is released
+      try {
+        this.txStateManager.setTxStatusFailed(txId, err)
```
Comment (danfinlay) [inline app/scripts/controllers/transactions/index.js]: Was `setTxStatusFailed()` throwing uncaught errors? Looking at it, it seems so straight forward. In fact, even its own internal `_setTxStatus(id, status)` method has its own try/catch.

File: app/scripts/controllers/transactions/nonce-tracker.js [hunk 3]
```diff
@@ -49,29 +49,35 @@ class NonceTracker {
...
+      nonceDetails.network = networkNonceResult
+
+      const nextNonce = Math.max(networkNonceResult.nonce, localNonceResult.nonce)
+      assert(Number.isInteger(nextNonce), `nonce-tracker - nextNonce is not an integer - got: (${typeof nextNonce}) "${nextNonce}"`)
+
+      // return nonce and release cb
+      return { nextNonce, nonceDetails, releaseLock }
+    } catch (err) {
+      // release lock if we encounter an error
+      releaseLock()
```
Comment (danfinlay) [inline app/scripts/controllers/transactions/nonce-tracker.js]: This looks like a great candidate for the `approved` bug, I'm so excited to see you find this.

File: app/scripts/metamask-controller.js [hunk 4]
```diff
@@ -466,7 +462,7 @@ module.exports = class MetamaskController extends EventEmitter {
    * @param  {} seed
    */
   async createNewVaultAndRestore (password, seed) {
-    const release = await this.createVaultMutex.acquire()
+    const releaseLock = await this.createVaultMutex.acquire()
```
Comment (danfinlay) [inline app/scripts/metamask-controller.js]: Very readable renaming.

**Review by danfinlay — DISMISSED** [review]
Looks good. I had some questions about one of the changes (txStateFailed threw errors?), but nothing that looks dangerous or bad. Has some important improvements that I'd love to ship asap.

**Comment by kumavis** [comment]
theres some style changes for consistency, the biggest change is the addition of 2 try-catchs where we were doing things that could error

File: app/scripts/controllers/transactions/index.js (hunk 2 above)
Comment (kumavis) [reply app/scripts/controllers/transactions/index.js]: i treated it as a black box and assumed errors could come from it
the errors could even be inside a eventemitter handler

File: app/scripts/metamask-controller.js [hunk 5]
```diff
@@ -436,28 +436,24 @@ module.exports = class MetamaskController extends EventEmitter {
    * @returns {Object} vault
    */
   async createNewVaultAndKeychain (password) {
-    const release = await this.createVaultMutex.acquire()
-    let vault
-
+    const releaseLock = await this.createVaultMutex.acquire()
```
Comment (bitpshr) [inline app/scripts/metamask-controller.js]: Can the mutex acquisition ever fail? Should we move this asynchronous operation inside the try / catch as well?

**Review by bitpshr — APPROVED** [review]
Looks good to me. I lack deep context here but the changes make sense if errors are being thrown from `setTxStatusFailed`, `getNonceLock`, and / or `createNewVaultAndKeychain`.

File: app/scripts/metamask-controller.js (hunk 5 above)
Comment (kumavis) [reply app/scripts/metamask-controller.js]: acquisition should never fail

**Pull request merged**